File: Binary_Search/704_binary_search/704.rs



fn search(nums: &Vec<i32>, target: &i32) -> i32 {
        
    fn recur(left: isize, right: isize, nums: &Vec<i32>, target: &i32) -> isize {
        
        if left > right {
            return -1;
        }
        
        let mid: isize = (left + right) / 2;
        
        if nums[mid as usize] == *target {
            return mid;
        }
        
        if nums[mid as usize] > *target {
            return recur(left, mid - 1, nums, target);
        }
        return recur(mid + 1, right, nums, target);
    }
    
    let right: isize = (nums.len() - 1) as isize;
    return recur(0, right, nums, target) as i32;   
}


fn main() {

    let nums: Vec<i32> = vec!(-1, 0, 3, 5, 9, 12);
    let target: i32 = 9;

    let result = search(&nums, &target);

    println!("nums : {:?} target : {} result : {}", nums, target, result);

}

File: Binary_Search/704_binary_search/704.py


def search(nums, target):
        
        left = 0
        right = len(nums) - 1
        
        def BinarySearch(nums, left, right):
            
            if left>right:
                return -1
            
            if nums[left] == target:
                return left
            elif nums[right] == target:
                return right
            
            mid = (left + right) // 2
            
            if nums[mid] == target:
                return mid
            
            if nums[mid] < target:
                return BinarySearch(nums, mid+1, right)
            else:
                return BinarySearch(nums, left, mid-1)
            
        return BinarySearch(nums, left, right)


nums = [-1, 0, 3, 5, 9, 12]
target = 9

print(f"nums : {nums}, target : {target}, result : {search(nums, target)}")